### ISSUE TITLE
ambiguity test hanging

### ISSUE THREAD

**Issue by colbec** [JuliaLang/julia#21160]
`make testall` after latest `git pull` results in a process that stops at testing `cmdlineargs` as in
```
...
error (2)                 |    1.12  |  0.00  |  0.0 | 11.33      | 1103.82 
cartesian (2)             |    0.11  |  0.02  | 17.8 |  1.14      | 1103.82 
asmvariant (2)            |    0.14  |  0.00  |  0.0 |  2.30      | 1103.82 
osutils (2)               |    0.02  |  0.00  |  0.0 |  0.23      | 1103.82 
channels (2)              |   10.55  |  0.70  |  6.6 | 428.04     | 1114.27 
iostream (2)              |    0.14  |  0.00  |  0.0 |  2.80      | 1114.27 
fft (2)                   |   19.29  |  0.87  |  4.5 | 723.95     | 1118.11 
dsp (2)                   |    6.26  |  0.24  |  3.8 | 266.36     | 1118.11 
examples (2)              |   13.79  |  0.47  |  3.4 | 648.76     | 1118.11 
distributed (4)           |   85.59  |  0.00  |  0.0 |  5.12      | 1003.38 
cmdlineargs (5)           |  111.27  |  0.07  |  0.1 | 65.80      | 952.91  
```
The system monitor shows there is no activity on any CPU apart from baseline. I force a halt with ctrl+c. Same thing repeated twice in a row reliably. and in each case cmdlineargs was running on a different worker.

In
```
Julia Version 0.6.0-pre.alpha.250
Commit 43fb9f0* (2017-03-25 05:22 UTC)
Platform Info:
  OS: Linux (x86_64-suse-linux)
  CPU: Intel(R) Core(TM) i5-4460  CPU @ 3.20GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)
```

**Comment by tkelman** [comment]
You'll notice some warnings `Skipping Base.<|` from having started the "ambiguous" test, but never a message that it finished. Caused by https://github.com/JuliaLang/julia/pull/20986#issuecomment-288885678

**Comment by tkelman** [comment]
I have a patch locally (will open a PR once it runs tests a few more times to see whether I can make the freeze happen) that reorders the ambiguity test to run first, which works around this for the purposes of avoiding the issue on CI and the buildbots, but doesn't address the underlying cause of whatever's making it freeze. I guess types introduced during some of the other tests are confusing the check.

**Comment by JeffBezanson** [comment]
I've been trying without luck to reproduce this, but it's notable that loading `TestHelpers` makes the ambiguity test *fail*, since it contains some ambiguous definitions.

I think `detect_ambiguities` might not be doing what we want with its module arguments. It's looking at all functions referenced in the specified modules, but it should perhaps only report results where all implicated *methods* are from the specified modules.

**Comment by tkelman** [comment]
Prior to #21161 I was able to get this to happen within a single-digit number of repeated runs of `make testall`. Though running tests in a single worker caused it to fail rather than hang as you said, due to ambiguities related to the Furlong type in TestHelpers.

**Comment by JeffBezanson** [comment]
Ah, that's a good clue. Could be related to #20230 --- unable to serialize the exception since it refers to stuff only defined on the worker.

**Comment by tkelman** [comment]
So - fix the existing ambiguities in TestHelpers and whatever other types get defined during tests (if this is worth the effort), try running a bunch of times with multiple workers or with a single worker with the ambiguity test running last, then call this a duplicate? Or is it a new problem that the serialization issue was resulting in a hang?

**Comment by StefanKarpinski** [comment]
Not a release blocker since it's not tanking CI. There's a bug here that seems to probably have to do with the test system and/or tasks. To address this: fix the hang, then revert the reordering here.

**Comment by vtjnash** [comment]
Probably was some sort of type-system intersection issue, hopefully fixed now, but since we merged  #21161 (mentioned above), it solved the test hanging issue. Given that we had no other reduction, there's nothing more we can do.
